plumbing/transport/server/loader.go

```diff
@@ -0,0 +1,51 @@
+package server
+
+import (
+	"gopkg.in/src-d/go-git.v4/plumbing/storer"
+	"gopkg.in/src-d/go-git.v4/plumbing/transport"
+	"gopkg.in/src-d/go-git.v4/storage/filesystem"
+
+	"srcd.works/go-billy.v1"
+	"srcd.works/go-billy.v1/os"
+)
+
+// DefaultLoader is a filesystem loader ignoring host and resolving paths to /.
+var DefaultLoader = NewFilesystemLoader(os.New("/"))
+
+// Loader loads repository's storer.Storer based on an optional host and a path.
+type Loader interface {
+	// Load loads a storer.Storer given a transport.Endpoint.
+	// Returns transport.ErrRepositoryNotFound if the repository does not
+	// exist.
+	Load(ep transport.Endpoint) (storer.Storer, error)
+}
+
+type fsLoader struct {
+	base billy.Filesystem
+}
+
+// NewFilesystemLoader creates a Loader that ignores host and resolves paths
+// with a given base filesystem.
+func NewFilesystemLoader(base billy.Filesystem) Loader {
+	return &fsLoader{base}
+}
+
+func (l *fsLoader) Load(ep transport.Endpoint) (storer.Storer, error) {
+	fs := l.base.Dir(ep.Path)
+	if _, err := fs.Stat("config"); err != nil {
+		return nil, transport.ErrRepositoryNotFound
+	}
+
+	return filesystem.NewStorage(fs)
+}
+
+type MapLoader map[transport.Endpoint]storer.Storer
```

Comment:
I've added further comments to explain each behaviour.

A Loader loads a repository (actually, a storer.Storer, which is a low-level representation of a repository) given an endpoint.

The fsLoader, as documented in NewFilesystemLoader, loads repositories from a file system. That is, it gets an endpoint's path, looks it up in the base file system, and if it's a repository, creates a storer for it. The DefaultLoader is a file system loader using the native OS filesystem.

MapLoader uses a lookup map to return storers that were already created by the user. This is useful for testing (e.g. registering a memory repository with a given path), but also for actual production applications, such as in-process pushes (i.e. source and destination repositories are both loaded in the same go-git process).

The relation in MapLoader is not bijective.

With respect to the name, feel free to suggest any alternative. `Load` takes an endpoint and it tries to build a storer for it if the repository exists.